### PR TITLE
Add Mask R-CNN

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR is a further development of the Mask R-CNN framework. Supersedes #22973.

Updates:

- improved variable names, docstrings (especially also for the configuration class)
- removed specific `__repr__` and `__nice__` methods
- favor ONNX-compatible code wherever possible, instead of if-else statements
- `torchvision` was already leveraged for NMS

Regarding this:

> In some of the model-side processing code, there's switching to CPU and casting back and forth between torch and numpy - it's not clear why.

=> this is because when placing masks on the GPU, this would cause OOM errors. Hence those are placed on the CPU.

### CODE REVIEW

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 1]
```diff
@@ -0,0 +1,3818 @@
...
+    avg_non_ignore=False,
+):
+    """Calculate the CrossEntropy loss.
+
+    Args:
+        pred (`torch.Tensor`):
+            The prediction with shape (N, C), C is the number of classes.
+        label (`torch.Tensor`):
+            The learning label of the prediction.
+        weight (`torch.Tensor`, *optional*):
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:255]:
```suggestion
        sample_weight (`torch.Tensor`, *optional*):
```

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 2]
```diff
@@ -0,0 +1,3818 @@
...
+    # average loss over non-ignored elements
+    # pytorch's official cross_entropy average loss over non-ignored elements
+    # refer to https://github.com/pytorch/pytorch/blob/56b43f4fec1f76953f15a627694d4bba34588969/torch/nn/functional.py#L2660  # noqa
+    if (avg_factor is None) and avg_non_ignore and reduction == "mean":
+        avg_factor = label.numel() - (label == ignore_index).sum().item()
+
+    # apply weights and do the reduction
+    if weight is not None:
+        weight = weight.float()
+    loss = weight_reduce_loss(loss, weight=weight, reduction=reduction, avg_factor=avg_factor)
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:285]: In stead of having this scale factor. We can:  
* Set `weight` to a vector of ones of size labels by default 
* mask out the weight, pred and label where label == ignore_index 
* pass those values to `nn.functional.cross_entropy`
* Average with the sample weights. 

This will reduce some of the if/else logic and make it clearer what's being averaged.

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 3]
```diff
@@ -0,0 +1,3818 @@
...
+            The model prediction, shape (N, num_class).
+        target (`torch.Tensor`):
+            The target of each prediction, shape (N, ).
+        topk (`int` or `Tuple[int]`, *optional*, defaults to 1):
+            If the predictions in `topk` matches the target, the predictions will be regarded as correct ones.
+        thresh (`float`, *optional*):
+            If not None, predictions with scores under this threshold are considered incorrect.
+
+    Returns:
+        `float` or `Tuple[float]`: If the input `topk` is a single integer,
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:201]: it's a list or float that's returned

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 4]
```diff
@@ -0,0 +1,3818 @@
...
+
+
+def multiclass_nms(
+    multi_bboxes, multi_scores, score_thr, nms_cfg, max_num=-1, score_factors=None, return_indices=False
+):
+    """NMS for multi-class bboxes.
+
+    Args:
+        multi_bboxes (`torch.Tensor`):
+            Shape (N, #class*4) or (N, 4) with N = number of objects.
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:3743]: Same with rest of docstring below 

```suggestion
            Shape (num_objects, num_classes*4) or (num_objects, 4)
```

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 5]
```diff
@@ -0,0 +1,3818 @@
...
+        self.backbone = AutoBackbone.from_config(config.backbone_config)
+        self.neck = MaskRCNNFPN(config, self.backbone.channels)
+        self.rpn_head = MaskRCNNRPN(config)
+        self.roi_head = MaskRCNNRoIHead(config)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def parse_losses(self, losses):
+        """Parse the raw outputs (losses) of the network.
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:3420]: Perhaps we could call this something like `aggregate_loss`? It's not really parsing anything, rather iterating and summing over dictionary values

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 6]
```diff
@@ -0,0 +1,3818 @@
...
+            if isinstance(loss_value, torch.Tensor):
+                log_vars[loss_name] = loss_value.mean()
+            elif isinstance(loss_value, list):
+                log_vars[loss_name] = sum(_loss.mean() for _loss in loss_value)
+            else:
+                raise TypeError(f"{loss_name} is not a tensor or list of tensors")
+
+        loss = sum(_value for _key, _value in log_vars.items() if "loss" in _key)
+
+        return loss
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:3443]: We can skip some the iteration logic

```suggestion
        loss = 0
        for loss_name, loss_value in losses.items():
            if "loss" not in loss_name: 
                continue
            if isinstance(loss_value, torch.Tensor):
                loss += loss_value.mean()
            elif isinstance(loss_value, list):
                for _loss in loss_value:
                    loss += _loss.mean()
            else:
                raise TypeError(f"{loss_name} is not a tensor or list of tensors")

        return loss
```

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 7]
```diff
@@ -0,0 +1,3818 @@
...
+        if self.with_mask:
+            mask_results = self._mask_forward_train(
+                feature_maps, sampling_results, bbox_results["bbox_feats"], gt_masks
+            )
+            losses.update(mask_results["loss_mask"])
+
+        return losses
+
+    def forward_test(self, hidden_states, proposal_list):
+        """Test without augmentation (originally called `simple_test`).
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:3322]: We should be careful with this as:

* We're not applying augmentation at any point, so this description could be confusing
* This method is used when there aren't any labels passed, but wouldn't be called if the model is in eval mode and labels were passed

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 8]
```diff
@@ -0,0 +1,3818 @@
...
+
+    def forward_test_mask(self, hidden_states, scale_factors, det_bboxes, rescale=True):
+        """Simple test for mask head without augmentation."""
+
+        # if det_bboxes is rescaled to the original image size, we need to
+        # rescale it back to the testing scale to obtain RoIs.
+        if rescale:
+            scale_factors = [scale_factor.to(det_bboxes[0].device) for scale_factor in scale_factors]
+        _bboxes = [
+            # det_bboxes[i][:, :4] * scale_factors[i] if rescale else det_bboxes[i][:, :4]
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:3244]: Maybe?

```suggestion
```

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 9]
```diff
@@ -0,0 +1,3818 @@
...
+
+        Removed with_shared_head here.
+        """
+        if not ((rois is not None) ^ (pos_indices is not None and bbox_feats is not None)):
+            raise ValueError("Either rois or (pos_indices and bbox_feats) should be specified")
+        if rois is not None:
+            mask_feats = self.mask_roi_extractor(x[: self.mask_roi_extractor.num_inputs], rois)
+        else:
+            if bbox_feats is None:
+                raise ValueError("bbox_feats must be specified when pos_indices is specified")
```
Comment (amyeroberts) [inline src/transformers/models/mask_rcnn/modeling_maskrcnn.py:3206]: This is already checked above

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this! 

I've done an initial review - commenting on the things that caught my eye. It wasn't however, extensive and I expect this PR will still need several rounds of review. 

General comments: 
* Could you add type hints to arguments in the functions? 
* Make sure all doc strings are filled out fully, including: descriptions of what the argument is, any shape information, shape information with explicit variable names 
* Given the complexity of this model, it needs a lot more tests beyond the defaults in the testing suite. In particular for implementations of loss averaging and NMS
*  `N` is used everywhere - it should be replaced with something more explicit
* As this PR is so necessarily big, I would try to find ways to make it as small as possible and then have follow up PRs for additional functionality. In particular, it would be a lot easier to review without the conditional onnx logic. Could you remove this and then add in a separate PR please?

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this! 

I've done an initial review - commenting on the things that caught my eye. It wasn't however, extensive and I expect this PR will still need several rounds of review. 

General comments: 
* Could you add type hints to arguments in the functions? 
* Make sure all doc strings are filled out fully, including: descriptions of what the argument is, any shape information, shape information with explicit variable names 
* Given the complexity of this model, it needs a lot more tests beyond the defaults in the testing suite. In particular for implementations of loss averaging and NMS
*  `N` is used everywhere - it should be replaced with something more explicit
* As this PR is so necessarily big, I would try to find ways to make it as small as possible and then have follow up PRs for additional functionality. In particular, it would be a lot easier to review without the conditional onnx logic. Could you remove this and then add in a separate PR please?

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/mask_rcnn/modeling_maskrcnn.py:261]: Would use the same term as PyTorch's one: https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 10]
```diff
@@ -0,0 +1,3818 @@
...
+                num_total_pos (`int`):
+                    Number of positive samples in all images.
+                num_total_neg (`int`):
+                    Number of negative samples in all images.
+        """
+        inside_flags = anchor_inside_flags(
+            flat_anchors, valid_flags, img_meta["img_shape"][-2:], self.train_cfg["allowed_border"]
+        )
+        if not inside_flags.any():
+            return (None,) * 7
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/mask_rcnn/modeling_maskrcnn.py:1929]: Because the function returns 7 elements.

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py [hunk 11]
```diff
@@ -0,0 +1,3818 @@
...
+
+
+class AssignResult:
+    """Stores assignments between predicted and truth boxes.
+
+    Attributes:
+        num_gts (`int`):
+            The number of truth boxes considered when computing this assignment
+        gt_indices (`torch.LongTensor`):
+            For each predicted box indicates the 1-based index of the assigned truth box. 0 means unassigned and -1
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/mask_rcnn/modeling_maskrcnn.py:602]: Because 0 means unassigned and -1 means ignore. This is why the indices start from 1 for positively assigned bounding boxes

File: src/transformers/models/mask_rcnn/modeling_maskrcnn.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/models/mask_rcnn/modeling_maskrcnn.py:210]: The function can also return a single float.
